### PR TITLE
go/services/regulated-assets-approval-server/internal/: Add kyc-status GET and DELETE methods

### PR DESCRIPTION
**Pull request by Zagan202**
<details>
  <summary>PR Checklist</summary>
  
### PR Structure

* [x] This PR has reasonably narrow scope (if not, break it down into smaller PRs).
* [x] This PR avoids mixing refactoring changes with feature changes (split into two PRs
  otherwise).
* [x] This PR's title starts with name of package that is most changed in the PR, ex.
  `services/friendbot`, or `all` or `doc` if the changes are broad or impact many
  packages.

### Thoroughness

* [x] This PR adds tests for the most critical parts of the new functionality or fixes.
* [x] I've updated any docs ([developer docs](https://www.stellar.org/developers/reference/), `.md`
  files, etc... affected by this change). Take a look in the `docs` folder for a given service,
  like [this one](https://github.com/stellar/go/tree/master/services/horizon/internal/docs).

### Release planning

* [ ] I've updated the relevant CHANGELOG ([here](services/horizon/CHANGELOG.md) for Horizon) if
  needed with deprecations, added features, breaking changes, and DB schema changes.
* [ ] I've decided if this PR requires a new major/minor version according to
  [semver](https://semver.org/), or if it's mainly a patch change. The PR is targeted at the next
  release branch if it's not a patch change.
</details>

### What

Add `kyc-status/` endpoint with the following methods
- [x] GET (detail)
  - Retrieve the user's kyc attempt(their row of the `accounts_kyc_status` table) 
- [x] DELETE
  - Remove the user's kyc attempt(their row of the `accounts_kyc_status` table) 

### Why

Have methods in place to help wallets check the status of a stellar account's KYC.
If approved they can retry POSTing their transaction to the server's `tx-approve/`  endpoint

### Known limitations

N/A

### CODE REVIEW

File: services/regulated-assets-approval-server/internal/serve/kyc-status/get_detail_handler.go [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+		log.Ctx(ctx).Debug("==== will log responses ====")
+		log.Ctx(ctx).Debugf("req: %+v", in)
+		log.Ctx(ctx).Debugf("resp: %+v", resp)
+		log.Ctx(ctx).Debugf("err: %+v", err)
+		log.Ctx(ctx).Debug("====  did log responses ====")
+	}()
+
+	// Check if getDetailRequest StellarAddressOrCallbackID value is present.
+	if in.StellarAddressOrCallbackID == "" {
+		return nil, httperror.NewHTTPError(http.StatusBadRequest, "Missing stellar address.")
```
Comment (marcelosalloum) [inline services/regulated-assets-approval-server/internal/serve/kyc-status/get_detail_handler.go:88]: 💡 The message only refers to stellar address but the variable could also contain a callback ID.

File: services/regulated-assets-approval-server/internal/serve/kyc-status/get_handler_test.go [hunk 2]
```diff
@@ -0,0 +1,25 @@
+package kycstatus
+
+import (
+	"testing"
+
+	"github.com/stellar/go/services/regulated-assets-approval-server/internal/db/dbtest"
+	"github.com/stretchr/testify/require"
+)
+
```
Comment (marcelosalloum) [inline services/regulated-assets-approval-server/internal/serve/kyc-status/get_handler_test.go:9]: Most of the GetDetailHandler functionality (including the core functionality) is not being tested in this file.

I realize you're already testing it in api_test.go but we usually make sure to cover the all edge cases in the unit tests files.

File: services/regulated-assets-approval-server/internal/serve/kyc-status/api_test.go [hunk 3]
```diff
@@ -199,3 +199,232 @@ func TestAPI_POSTKYCStatus(t *testing.T) {
...
+	resp = w.Result()
+	assert.Equal(t, http.StatusNotFound, resp.StatusCode)
+	body, err = ioutil.ReadAll(resp.Body)
+	require.NoError(t, err)
+
+	// TEST ok response for deleteing account's kyc record from db's "accounts_kyc_status" table.
+	wantDeleteResponseNotFound := `{
+		"error": "Not found."
+		}`
+	require.JSONEq(t, wantDeleteResponseNotFound, string(body))
```
Comment (marcelosalloum) [inline services/regulated-assets-approval-server/internal/serve/kyc-status/api_test.go:429]: This comment description is not correct.

File: services/regulated-assets-approval-server/internal/serve/kyc-status/api_test.go [hunk 4]
```diff
@@ -199,3 +199,232 @@ func TestAPI_POSTKYCStatus(t *testing.T) {
 			}`
 	require.JSONEq(t, wantPostResponseNotFound, string(body))
 }
```
Comment (marcelosalloum) [inline services/regulated-assets-approval-server/internal/serve/kyc-status/api_test.go:201]: Not a change for this PR, but I just realized the api_test.go file is inside the kyc-status package. The best way to test the API though is launching the API handler in the serve package. For this project, the concept of API only exists in the serve package, that's why we should test the API there and do only unit tests in the kyc-status package.

The tests you added look good though, so no need to remove any of them. You can probably reorganize these API tests in a future PR.

**Review by marcelosalloum — COMMENTED** [review]
Added a few suggestions 💡 below. I also marked with ⁉️  a regression issue regarding [go tests naming conventions](https://ieftimov.com/post/testing-in-go-naming-conventions/).

File: services/regulated-assets-approval-server/internal/serve/kyc-status/get_handler_test.go [hunk 5]
```diff
@@ -0,0 +1,25 @@
+package kycstatus
+
```
[the remark this answers is not in the document]
Comment (Zagan202) [reply services/regulated-assets-approval-server/internal/serve/kyc-status/get_handler_test.go:2]: renamed to `get_detail_handler_test.go` here eff4146adef1eaf438b035c55ee7d8bae86b4d5d

File: services/regulated-assets-approval-server/README.md [hunk 6]
```diff
@@ -202,7 +204,38 @@ If their KYC was rejected they should see a rejection response.
...
+**Response:**
+
+```json
+{
+  "stellar_address": "GA2DMTP67JT4LQ4CFTUONFBFGKPO6VONW4LWJNOIY2WPRJLUV44MJZOK",
+  "email_address": "test@test.com",
+  "created_at": "2021-03-26T09:35:06.907293-03:00",
+  "kyc_submitted_at": "2021-03-26T14:03:43.314334-03:00",
+  "approved_at": "2021-03-26T14:03:43.314334-03:00",
+  "rejected_at": ""
```
[the remark this answers is not in the document]
Comment (Zagan202) [reply services/regulated-assets-approval-server/README.md]: Good catch, I added and edited the examples here to be reflected of what the response will return.
This includes if the kyc isnt submitted, was approved, or rejected.
2d0725048e82f1f03ad97c091951d0b7f8aba0e4

File: services/regulated-assets-approval-server/internal/serve/kyc-status/api_test.go (hunk 4 above)
Comment (Zagan202) [reply services/regulated-assets-approval-server/internal/serve/kyc-status/api_test.go:201]: Sounds like a good plan I made an issue for this [here](https://github.com/stellar/nebula/issues/196)

File: services/regulated-assets-approval-server/internal/serve/kyc-status/get_handler_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (Zagan202) [reply services/regulated-assets-approval-server/internal/serve/kyc-status/get_handler_test.go]: renamed to [get_detail_handler_test.go](https://github.com/stellar/go/blob/23e820c1bc3cf2de3394f6d26652ca667bb9859f/services/regulated-assets-approval-server/internal/serve/kyc-status/get_detail_handler_test.go)

File: services/regulated-assets-approval-server/internal/serve/kyc-status/api_test.go (hunk 3 above)
Comment (Zagan202) [reply services/regulated-assets-approval-server/internal/serve/kyc-status/api_test.go]: Changed the comment [here](d895b88)

File: services/regulated-assets-approval-server/internal/serve/kyc-status/get_detail_handler.go (hunk 1 above)
Comment (Zagan202) [reply services/regulated-assets-approval-server/internal/serve/kyc-status/get_detail_handler.go]: Changed error response here 1134353

File: services/regulated-assets-approval-server/internal/serve/kyc-status/get_handler_test.go (hunk 2 above)
Comment (Zagan202) [reply services/regulated-assets-approval-server/internal/serve/kyc-status/get_handler_test.go]: I've added tests to these files
- [delete_handler_test.go](https://github.com/stellar/go/blob/64825d4aca32f14233bc9aae99b2c81eaa4f6b86/services/regulated-assets-approval-server/internal/serve/kyc-status/delete_handler_test.go) 
- [get_detail_handler_test.go](https://github.com/stellar/go/blob/64825d4aca32f14233bc9aae99b2c81eaa4f6b86/services/regulated-assets-approval-server/internal/serve/kyc-status/get_detail_handler_test.go)

Please let me know if this is what we're looking for, 🙏🏾

File: services/regulated-assets-approval-server/internal/serve/kyc-status/api_test.go [hunk 7]
```diff
@@ -175,7 +175,7 @@ func TestAPI_POSTKYCStatus(t *testing.T) {
 
 	// TEST "Missing email_address" error response.
 	wantPostResponseMissingEmail := `{
-		"error": "Missing email_address."
+		"error": "missing email_address"
 	}`
```
Comment (marcelosalloum) [inline services/regulated-assets-approval-server/internal/serve/kyc-status/api_test.go:179]: 👀  This is not an internal Go error but rather an API error that can be shown to the user, so it's better to keep it in a more readable form like it was before.

Also, I noticed you changed it in a few other tests as well, could you revert those changes too?

File: services/regulated-assets-approval-server/internal/serve/kyc-status/api_test.go [hunk 8]
```diff
@@ -193,9 +193,236 @@ func TestAPI_POSTKYCStatus(t *testing.T) {
 	body, err = ioutil.ReadAll(resp.Body)
 	require.NoError(t, err)
 
-	// TEST "Not Found" error response.
+	// TEST "not found" error response.
 	wantPostResponseNotFound := `{
-			"error": "Not found."
+			"error": "not found"
 			}`
```
Comment (marcelosalloum) [inline services/regulated-assets-approval-server/internal/serve/kyc-status/api_test.go:199]: Same here about "API user facing error" vs "internal Go error".

File: services/regulated-assets-approval-server/internal/serve/kyc-status/delete_handler_test.go [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+	existQuery := `
+		SELECT EXISTS(
+			SELECT stellar_address
+			FROM accounts_kyc_status
+			WHERE stellar_address = $1
+		)`
+	var exists bool
+	err = h.DB.QueryRowContext(ctx, existQuery, accountKP.Address()).Scan(&exists)
+	require.NoError(t, err)
+	assert.True(t, exists)
```
Comment (marcelosalloum) [inline services/regulated-assets-approval-server/internal/serve/kyc-status/delete_handler_test.go:76]: 💡 You've just executed an INSERT query and you validated it went correctly, so this chunk of code is redundant and can be removed.

File: services/regulated-assets-approval-server/internal/serve/kyc-status/get_detail_handler_test.go [hunk 10]
```diff
@@ -0,0 +1,132 @@
...
+	nullTimePtr.Valid = true
+	timeNow := time.Now()
+	nullTimePtr.Time = timeNow
+
+	// Send a valid Pointer to timePointerIfValid.
+	// TEST if timePointer is valid and if return a time.Time pointer equals the time.Now().
+	timePointer = timePointerIfValid(nullTimePtr)
+	require.NotNil(t, timePointer)
+	assert.Equal(t, &timeNow, timePointer)
+}
```
Comment (marcelosalloum) [inline services/regulated-assets-approval-server/internal/serve/kyc-status/get_detail_handler_test.go:50]: Cool, good job adding a test for this method! 🙌

File: services/regulated-assets-approval-server/internal/serve/kyc-status/get_detail_handler_test.go [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+	wantKycGetResponse := kycGetResponse{
+		StellarAddress: accountKP.Address(),
+		CallbackID:     callbackID,
+		EmailAddress:   emailAddress,
+		CreatedAt:      kycGetResp.CreatedAt,
+		KYCSubmittedAt: kycGetResp.KYCSubmittedAt,
+		ApprovedAt:     kycGetResp.ApprovedAt,
+		RejectedAt:     kycGetResp.RejectedAt,
+	}
+	assert.Equal(t, &wantKycGetResponse, kycGetResp)
```
Comment (marcelosalloum) [inline services/regulated-assets-approval-server/internal/serve/kyc-status/get_detail_handler_test.go:125]: 💡 Could you also test if the account is retrievable using a callback ID?

**Review by marcelosalloum — COMMENTED** [review]
Almost there, added a few more suggestions below:

File: services/regulated-assets-approval-server/internal/serve/kyc-status/api_test.go (hunk 7 above)
Comment (Zagan202) [reply services/regulated-assets-approval-server/internal/serve/kyc-status/api_test.go]: I took note of this tip and changed the external errors' formatting.

File: services/regulated-assets-approval-server/internal/serve/kyc-status/delete_handler_test.go (hunk 9 above)
Comment (Zagan202) [reply services/regulated-assets-approval-server/internal/serve/kyc-status/delete_handler_test.go]: Noted, should be removed now.

File: services/regulated-assets-approval-server/internal/serve/kyc-status/get_detail_handler_test.go (hunk 11 above)
Comment (Zagan202) [reply services/regulated-assets-approval-server/internal/serve/kyc-status/get_detail_handler_test.go]: Test was added in this [commit](19770dd)

File: services/regulated-assets-approval-server/internal/serve/kyc-status/get_detail_handler_test.go [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+
+	// TEST if response timestamps are present or null.
+	require.NotNil(t, kycGetResp.CreatedAt)
+	require.NotNil(t, kycGetResp.KYCSubmittedAt)
+	require.NotNil(t, kycGetResp.ApprovedAt)
+	require.Nil(t, kycGetResp.RejectedAt)
+
+	/// Prepare and send getDetailRequest to an account in the db; using stellar address.
```
Comment (marcelosalloum) [inline services/regulated-assets-approval-server/internal/serve/kyc-status/get_detail_handler_test.go:122]: 💡 One of the bad things about comments is it's easy to forget to update them when you copy/paste code:
```suggestion
	// Prepare and send getDetailRequest to an account in the db; using callbackID.
```

**Review by marcelosalloum — COMMENTED** [review]
I noticed some exaggeration in the usage of comments in `TestGetDetailHandlerHandle`, this large amount of comments becomes noisy when reading the code, don't you think? Would you mind trying to organize it a little better?

I'm not against commenting in the code as long as it brings value. Here are a few quick reads on the subject, do you mind taking a look?
* [Comments: the good, the bad, and the ugly](https://www.freecodecamp.org/news/code-comments-the-good-the-bad-and-the-ugly-be9cc65fbf83/)
* [Stackoverflow thread](https://softwareengineering.stackexchange.com/questions/1/comments-are-a-code-smell)
* [Don't comment your code (Clean Code)](https://devdeejay.medium.com/dont-comment-your-code-rewrite-it-a145d655f87b)

File: services/regulated-assets-approval-server/internal/serve/kyc-status/get_detail_handler_test.go (hunk 12 above)
Comment (Zagan202) [reply services/regulated-assets-approval-server/internal/serve/kyc-status/get_detail_handler_test.go]: Addressed here after reorganization b387074

File: services/regulated-assets-approval-server/internal/serve/kyc-status/get_detail_handler_test.go [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+	require.EqualError(t, err, "Not found.")
+
+	// Prepare and send getDetailRequest to a callbackID not in the db.
+	callbackID := uuid.New().String()
+	in = getDetailRequest{StellarAddressOrCallbackID: callbackID}
+	kycGetResp, err = h.handle(ctx, in)
+
+	// TEST error "Not found.".
+	require.Nil(t, kycGetResp)
+	require.EqualError(t, err, "Not found.")
```
[the remark this answers is not in the document]
Comment (Zagan202) [reply services/regulated-assets-approval-server/internal/serve/kyc-status/get_detail_handler_test.go]: Fixed code organization here 3bdfaeb

**Review by marcelosalloum — APPROVED** [review]
LGTM!
